firedrake/function.py

```diff
@@ -487,3 +500,197 @@ def __idiv__(self, expr):
             assemble_expressions.IDiv(self, expr))
 
         return self
+
+    @utils.cached_property
+    def ctypes(self):
+        # Retrieve data from Python object
+        function_space = self.function_space()
+        mesh = function_space.mesh()
+        coordinates = mesh.coordinates
+        coordinates_space = coordinates.function_space()
+
+        # Store data into ``C struct''
+        c_function = _CFunction()
+        c_function.n_cols = mesh.num_cells()
+        if hasattr(mesh, '_layers'):
+            c_function.n_layers = mesh.layers - 1
+        else:
+            c_function.n_layers = 1
+        c_function.coords = coordinates.dat.data.ctypes.data_as(POINTER(c_double))
+        c_function.coords_map = coordinates_space.cell_node_list.ctypes.data_as(POINTER(c_int))
+        c_function.f = self.dat.data.ctypes.data_as(POINTER(c_double))
+        c_function.f_map = function_space.cell_node_list.ctypes.data_as(POINTER(c_int))
+        c_function.sidx = mesh.spatial_index and mesh.spatial_index.ctypes
+
+        # Return pointer
+        return ctypes.pointer(c_function)
+
+    @utils.cached_property
+    def _c_evaluate(self):
+        result = make_c_evaluate(self)
+        result.restype = int
+        return result
+
+    def evaluate(self, coord, mapping, component, index_values):
+        # Called by UFL when evaluating expressions at coordinates
+        return self.at(coord)
```

Comment:
I don't really what these parameters mean, haven't quite found the documentation in UFL. But I can raise an exception is they are specified.